### PR TITLE
Fix msgdispatcher concurrency issue and add prometheus metrics

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

issues: https://github.com/milvus-io/milvus/issues/22166

### CODE REVIEW

**Comment by bigsheeper** [comment]
rerun ut

File: internal/mq/msgdispatcher/manager.go [hunk 1]
```diff
@@ -150,14 +154,14 @@ func (c *dispatcherManager) Run() {
...
 	log.Info("dispatcherManager is running...")
-	ticker := time.NewTicker(CheckPeriod)
-	defer ticker.Stop()
 	for {
 		select {
 		case <-c.closeChan:
 			log.Info("dispatcherManager exited")
 			return
-		case <-ticker.C:
+		case <-time.NewTicker(10 * time.Second).C:
```
Comment (czs007) [inline internal/mq/msgdispatcher/manager.go:162]: may be resource leak ?

**Comment by bigsheeper** [comment]
image

**Comment by bigsheeper** [comment]
image

**Comment by bigsheeper** [comment]
rerun ut

**Comment by bigsheeper** [comment]
rerun ut

File: internal/mq/msgdispatcher/dispatcher.go [hunk 2]
```diff
@@ -192,39 +200,22 @@ func (d *Dispatcher) work() {
...
+			targetPacks := d.groupingMsgs(pack)
+			for vchannel, p := range targetPacks {
+				var err error
+				var t = d.targets[vchannel]
+				if d.isMain {
+					err = t.send(p)
+				} else {
+					for {
+						err = t.send(p)
+						if err == nil || funcutil.CheckCtxValid(d.ctx) {
```
Comment (czs007) [inline internal/mq/msgdispatcher/dispatcher.go:212]: !funcutil.CheckCtxValid(d.ctx)

File: internal/mq/msgdispatcher/dispatcher.go [hunk 3]
```diff
@@ -192,39 +200,22 @@ func (d *Dispatcher) work() {
...
-					for k := range targetPacks {
-						targetPacks[k].Msgs = append(targetPacks[k].Msgs, msg)
+			targetPacks := d.groupingMsgs(pack)
+			for vchannel, p := range targetPacks {
+				var err error
+				var t = d.targets[vchannel]
+				if d.isMain {
+					err = t.send(p)
+				} else {
+					for {
```
Comment (czs007) [inline internal/mq/msgdispatcher/dispatcher.go:210]: please add comment to explain this for loop.

**Pull request merged**